code/game/gamemodes/wizard/soulstone.dm

```diff
@@ -106,69 +107,278 @@
 ////////////////////////////Proc for moving soul in and out off stone//////////////////////////////////////
 
 
+/obj/item/proc/capture_soul(var/target, var/mob/user as mob)
+	if(istype(target, /mob/living/carbon/human))
+		var/mob/living/carbon/human/T = target
+		//first of all, let's check that our target has a soul, somewhere
+
+		if(!T.client)
+			//no client? the target could be either braindead, decapitated, or catatonic, let's check which
+			var/datum/organ/internal/brain/B = T.internal_organs_by_name["brain"]
+			if(B)
+				//our target has their head, did they have a soul in the first place? if so, where is it right now
+				if(!T.mind)
+					//if a mob doesn't have a mind, that means it never had a player controlling him
+					to_chat(user, "<span class='warning'>The soul stone isn't reacting, looks like this target doesn't have much of a soul.</span>")
+					return
+				else
+					//otherwise, that means the player either disconnected or ghosted. we can track their key from their mind,
+					//but first let's make sure that they are dead or in crit
+					var/mob/new_target = null
+					for(var/mob/M in player_list)
+						if(M.key == T.mind.key)
+							new_target = M
+					if(!new_target)
+						to_chat(user, "<span class='warning'>The soul stone isn't reacting, looks like this target's soul went far, far away.</span>")
+						return
+					else if(!istype(new_target,/mob/dead/observer))
+						to_chat(user, "<span class='warning'>The soul stone isn't reacting, looks like this target's soul already reincarnated.</span>")
+						return
+					else
+						to_chat(new_target, "<span class='danger'>You feel your soul getting sucked into the soul stone.</span>")
+						to_chat(user, "<span class='rose'>The soul stone reacts to the corpse and starts glowing.</span>")
+						capture_soul_process(user,new_target.client,T)
+			else
+				//aw shit, let's try and locate the head.
```

Comment:
You can chop this proc in half by moving all human relevant code into just this section